src/NRules/NRules.Fluent/RuleDefinitionFactory.cs

```diff
@@ -36,11 +36,22 @@ public IRuleDefinition Create(Rule rule)
     {
         try
         {
-            return rule.GetDefinition();
+            return BuildDefinition(rule);
         }
         catch (Exception e)
         {
             throw new RuleDefinitionException("Failed to build rule definition", rule.GetType(), e);
         }
     }
+
+    private IRuleDefinition BuildDefinition(Rule rule)
+    {
+        var metadata = new RuleMetadata(rule.GetType());
+
+        var context = RuleBuildContext.CreateNew(metadata);
+
+        rule.Define(context);
```

Comment:
@gritcsenko I would instantiate the builder here, in the factory, and just pass to Rule.Define(builder). Then RuleBuilderContext is just an internal detail of the Rule class. The purpose of the RuleBuilderContext is mostly to wrap RuleBuilder and SymbolStack and it has no use outside of the Rule

Reply:
Struct moved